### PR TITLE
Login and Unix group names check via the perun-namespaces.properties file

### PR DESCRIPTION
**Pull request by Holdo**
Task #2580 - Login and Unix group names check via the perun-namespaces.properties file

### CODE REVIEW

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java [hunk 1]
```diff
@@ -579,7 +594,29 @@ public void checkFormatOfShell(String shell, Attribute attribute) throws WrongAt
 		if (!match.matches()) {
 			throw new WrongAttributeValueException(attribute, "Bad shell attribute format " + shell);
 		}
+	}
 
+	public void checkAttributeRegex(User user, Attribute attribute, String defaultRegex) throws WrongAttributeValueException {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java]: Why user there?

**Comment by Holdo** [comment]
Not finished yet

**Comment by Holdo** [comment]
Finished

**Comment by stavamichal** [comment]
For info: not founded file with namespace mean log error which can fill log with lots of error spam with no importance

**Comment by stavamichal** [comment]
test this please

**Comment by stavamichal** [comment]
test this please

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java [hunk 2]
```diff
@@ -481,14 +486,40 @@ public boolean haveRightToWriteAttributeInAnyGroupOrResource(PerunSessionImpl se
...
+		}
 	}
 
-	public void checkUnpermittedUserLogins(Attribute loginAttribute) throws WrongAttributeValueException {
+	public void checkUnpermittedUserLogins(Attribute loginAttribute) throws InternalErrorException, WrongAttributeValueException {
 		if(loginAttribute == null) return;
-		if(unpermittedNamesForUserLogins.contains(loginAttribute.getValue())) throw new WrongAttributeValueException(loginAttribute, "This login is not permitted.");
+
+		if (perunNamespaces == null) {
+			perunNamespaces = BeansUtils.getAllPropertiesFromCustomConfiguration("perun-namespaces.properties");
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java]: If file not found exception is thrown (when file not exists, correct status), Internal Error Exception is thrown and everything fail. In this case we need to set perunNamespaces map to empty one and continue.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java [hunk 3]
```diff
@@ -481,14 +486,40 @@ public boolean haveRightToWriteAttributeInAnyGroupOrResource(PerunSessionImpl se
 		return groupNameNamespaces;
 	}
 
-	public void checkReservedUnixGroupNames(Attribute groupNameAttribute) throws WrongAttributeValueException {
+	public void checkReservedUnixGroupNames(Attribute groupNameAttribute) throws InternalErrorException, WrongAttributeValueException {
 		if(groupNameAttribute == null) return;
-		if(reservedNamesForUnixGroups.contains(groupNameAttribute.getValue())) throw new WrongAttributeValueException(groupNameAttribute, "This groupName is reserved.");
+
+		if (perunNamespaces == null) {
+			perunNamespaces = BeansUtils.getAllPropertiesFromCustomConfiguration("perun-namespaces.properties");
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java]: If file not found exception is thrown (when file not exists, correct status), Internal Error Exception is thrown and everything fail. In this case we need to set perunNamespaces map to empty one and continue.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java [hunk 4]
```diff
@@ -579,7 +610,30 @@ public void checkFormatOfShell(String shell, Attribute attribute) throws WrongAt
 		if (!match.matches()) {
 			throw new WrongAttributeValueException(attribute, "Bad shell attribute format " + shell);
 		}
+	}
 
+	public void checkAttributeRegex(Attribute attribute, String defaultRegex) throws InternalErrorException, WrongAttributeValueException {
+		if (attribute == null || attribute.getValue() == null) throw new InternalErrorException("Attribute or it's value is null.");
+		String attributeValue = (String) attribute.getValue();
+		try {
+			String regex = BeansUtils.getPropertyFromCustomConfiguration("perun-namespaces.properties", attribute.getFriendlyName() + ":regex");
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java]: This is the same problem like with logins and group names. We don't want to read the file for every time we call this method. Better will be fill or use perunNamespaces map where this value should be too.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java [hunk 5]
```diff
@@ -481,14 +486,40 @@ public boolean haveRightToWriteAttributeInAnyGroupOrResource(PerunSessionImpl se
...
-	public void checkUnpermittedUserLogins(Attribute loginAttribute) throws WrongAttributeValueException {
+	public void checkUnpermittedUserLogins(Attribute loginAttribute) throws InternalErrorException, WrongAttributeValueException {
 		if(loginAttribute == null) return;
-		if(unpermittedNamesForUserLogins.contains(loginAttribute.getValue())) throw new WrongAttributeValueException(loginAttribute, "This login is not permitted.");
+
+		if (perunNamespaces == null) {
+			perunNamespaces = BeansUtils.getAllPropertiesFromCustomConfiguration("perun-namespaces.properties");
+		}
+		String unpermittedNames;
+		if ((unpermittedNames = perunNamespaces.get(loginAttribute.getFriendlyName() + ":reservedNames")) != null) {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java]: For better readability please fill it when declare variable and in the if clause just test if it is not null.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java [hunk 6]
```diff
@@ -481,14 +486,40 @@ public boolean haveRightToWriteAttributeInAnyGroupOrResource(PerunSessionImpl se
...
-	public void checkReservedUnixGroupNames(Attribute groupNameAttribute) throws WrongAttributeValueException {
+	public void checkReservedUnixGroupNames(Attribute groupNameAttribute) throws InternalErrorException, WrongAttributeValueException {
 		if(groupNameAttribute == null) return;
-		if(reservedNamesForUnixGroups.contains(groupNameAttribute.getValue())) throw new WrongAttributeValueException(groupNameAttribute, "This groupName is reserved.");
+
+		if (perunNamespaces == null) {
+			perunNamespaces = BeansUtils.getAllPropertiesFromCustomConfiguration("perun-namespaces.properties");
+		}
+		String reservedNames;
+		if ((reservedNames = perunNamespaces.get(groupNameAttribute.getFriendlyName() + ":reservedNames")) != null) {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java]: For better readability please fill it when declare variable and in the if clause just test if it is not null.

**Comment by stavamichal** [comment]
test this please

**Comment by stavamichal** [comment]
test this please

**Comment by stavamichal** [comment]
test this please

**Comment by stavamichal** [comment]
test this please

**Comment by stavamichal** [comment]
This one is prepared to be merged.

**Pull request merged**